From 475b360c0520447da77d664e7fd47aaf3e319d00 Mon Sep 17 00:00:00 2001
From: denglei <denglei@eswincomputing.com>
Date: Thu, 10 Apr 2025 09:56:22 +0800
Subject: [PATCH 466/652] fix:Resolve the issue of system crashes.

Changelogs:
	Resolve the issue of system crashes caused by DSP driver insmod/uninsomd

Signed-off-by: denglei <denglei@eswincomputing.com>
---
 drivers/soc/eswin/ai_driver/dsp/dsp_main.c | 17 +++++++++--------
 drivers/soc/eswin/ai_driver/dsp/dsp_main.h |  3 +++
 2 files changed, 12 insertions(+), 8 deletions(-)

diff --git a/drivers/soc/eswin/ai_driver/dsp/dsp_main.c b/drivers/soc/eswin/ai_driver/dsp/dsp_main.c
index 5ad8b8471f82..a5cc197fe138 100644
--- a/drivers/soc/eswin/ai_driver/dsp/dsp_main.c
+++ b/drivers/soc/eswin/ai_driver/dsp/dsp_main.c
@@ -949,9 +949,6 @@ static int es_dsp_hw_probe(struct platform_device *pdev)
 	int ret;
 	char nodename[sizeof("es-dsp") + 3 * sizeof(int)];
 	struct es_dsp *dsp;
-#if defined(CONFIG_PM_DEVFREQ)
-	struct devfreq *df;
-#endif
 
 	dsp = devm_kzalloc(&pdev->dev,
 			   sizeof(*dsp) + sizeof(struct es_dsp_stats) +
@@ -1028,15 +1025,15 @@ static int es_dsp_hw_probe(struct platform_device *pdev)
 		goto err_dsp_devfreq;
 	}
 
-	df = devm_devfreq_add_device(&pdev->dev, &dsp_devfreq_profile, DEVFREQ_GOV_SIMPLE_ONDEMAND, &ondemand_data);
-	if (IS_ERR(df)) {
+	dsp->df = devm_devfreq_add_device(&pdev->dev, &dsp_devfreq_profile, DEVFREQ_GOV_SIMPLE_ONDEMAND, &ondemand_data);
+	if (IS_ERR(dsp->df)) {
 		dsp_err("%s, %d, add devfreq failed\n", __func__, __LINE__);
-		ret = PTR_ERR(df);
+		ret = PTR_ERR(dsp->df);
 		goto err_dsp_devfreq;
 	};
 
 	/* Register opp_notifier to catch the change of OPP  ????*/
-	ret = devm_devfreq_register_opp_notifier(&pdev->dev, df);
+	ret = devm_devfreq_register_opp_notifier(&pdev->dev, dsp->df);
 	if (ret < 0) {
 		dev_err(&pdev->dev, "failed to register opp notifier\n");
 		return ret;
@@ -1105,7 +1102,7 @@ static int es_dsp_hw_probe(struct platform_device *pdev)
 	es_dsp_clk_disable(dsp);
 err_dsp_clk:
 #if defined(CONFIG_PM_DEVFREQ)
-	devm_devfreq_remove_device(dsp->dev, df);
+	devm_devfreq_remove_device(dsp->dev, dsp->df);
 err_dsp_devfreq:
 #endif
 	dsp_disable_mbox_clock(dsp);
@@ -1151,6 +1148,10 @@ static int es_dsp_hw_remove(struct platform_device *pdev)
 
 	win2030_tbu_power(dsp->dev, false);
 
+#if defined(CONFIG_PM_DEVFREQ)
+	devm_devfreq_remove_device(dsp->dev, dsp->df);
+#endif
+
 	es_dsp_clk_disable(dsp);
 	dsp_disable_mbox_clock(dsp);
 	es_dsp_unmap_resource(dsp);
diff --git a/drivers/soc/eswin/ai_driver/dsp/dsp_main.h b/drivers/soc/eswin/ai_driver/dsp/dsp_main.h
index f98ae647cd7c..99aab26af20f 100644
--- a/drivers/soc/eswin/ai_driver/dsp/dsp_main.h
+++ b/drivers/soc/eswin/ai_driver/dsp/dsp_main.h
@@ -191,6 +191,9 @@ struct es_dsp {
 	dsp_kmd_perf_t op_perf[MAX_DSP_TASKS];
 	dsp_fw_perf_t op_fw_perf[MAX_DSP_TASKS];
 	atomic_t dmabuf_mapped_cnt;
+#if defined(CONFIG_PM_DEVFREQ)
+	struct devfreq *df;
+#endif
 };
 
 #define DSP_FIRMWARE_IOVA 0xfe000000
-- 
2.54.0

